File: src/glottisdale/collage/align.py
"""Aligner interface and backends."""

import logging
import shutil
from abc import ABC, abstractmethod
from pathlib import Path

from glottisdale.types import Syllable
from glottisdale.collage.transcribe import transcribe
from glottisdale.collage.syllabify import syllabify_words

logger = logging.getLogger(__name__)


class Aligner(ABC):
    """Abstract base for speech alignment backends."""

    @abstractmethod
    def process(self, audio_path: Path) -> dict:
        """Transcribe and align audio, returning syllable-level timestamps.

        Returns:
            Dict with keys:
                text: Full transcript
                words: List of word dicts with timestamps
                syllables: List of Syllable objects
        """


class DefaultAligner(Aligner):
    """Whisper ASR + g2p_en + ARPABET syllabifier.

    Word-level timestamps from Whisper, phoneme conversion via g2p_en,
    syllable timing estimated by proportional distribution.
    """

    def __init__(self, whisper_model: str = "base", language: str = "en", **kwargs):
        self.whisper_model = whisper_model
        self.language = language

    def process(self, audio_path: Path) -> dict:
        result = transcribe(audio_path, model_name=self.whisper_model, language=self.language)
        syllables = syllabify_words(result["words"])
        return {
            "text": result["text"],
            "words": result["words"],
            "syllables": syllables,
        }


def _get_bfa_class():
    """Lazy import of BFAAligner to avoid hard dependency."""
    from glottisdale.collage.bfa import BFAAligner
    return BFAAligner


def _bfa_available() -> bool:
    """Check if BFA and espeak-ng are both available."""
    try:
        import bournemouth_aligner  # noqa: F401
    except ImportError:
        return False
    if shutil.which("espeak-ng") is None:
        return False
    return True


# Registry of available backends (lazy import for BFA)
_ALIGNERS = {
    "default": DefaultAligner,
    "bfa": _get_bfa_class,
}


def get_aligner(name: str, **kwargs) -> Aligner:
    """Get an aligner backend by name.

    Modes:
        "default" — Whisper + g2p_en + ARPABET proportional timing.
        "bfa" — Whisper + BFA phoneme-level alignment (requires bournemouth-forced-aligner + espeak-ng).
        "auto" — Tries BFA first, falls back to default if unavailable.
    """
    if name == "auto":
        if _bfa_available():
            logger.info("Auto-detected BFA + espeak-ng, using BFA aligner")
            return _get_bfa_class()(**kwargs)
        else:
            logger.info("BFA not available, falling back to default (proportional) aligner")
            return DefaultAligner(**kwargs)

    if name not in _ALIGNERS:
        raise ValueError(
            f"Unknown aligner: {name!r}. Available: {list(_ALIGNERS.keys()) + ['auto']}"
        )

    factory = _ALIGNERS[name]
    # BFA entry is a function returning the class (lazy import)
    if name == "bfa":
        try:
            cls = factory()
        except ImportError as e:
            raise ImportError(
                f"BFA aligner requires 'bournemouth-forced-aligner' package: {e}"
            ) from e
        return cls(**kwargs)

    return factory(**kwargs)

File: crates/core/src/language/align.rs
//! Aligner interface and backends.
//!
//! Provides syllable-level timestamps from audio files using different
//! alignment strategies:
//! - DefaultAligner: Whisper ASR + G2P + ARPABET syllabifier
//! - BfaAligner: Planned native forced alignment (see issue #21)

use std::path::Path;

use anyhow::{Result, bail};

use crate::types::AlignmentResult;

use super::syllabify;
use super::transcribe;

/// Alignment backend trait.
pub trait Aligner: Send + Sync {
    /// Backend name for caching/display.
    fn name(&self) -> &str;

    /// Transcribe and align audio, returning syllable-level timestamps.
    fn process(
        &self,
        audio_path: &Path,
        model_dir: Option<&Path>,
    ) -> Result<AlignmentResult>;
}

/// Whisper ASR + G2P + ARPABET syllabifier.
///
/// Word-level timestamps from Whisper, phoneme conversion via CMU dict,
/// syllable timing estimated by proportional distribution.
pub struct DefaultAligner {
    pub whisper_model: String,
    pub language: String,
}

impl DefaultAligner {
    pub fn new(whisper_model: &str, language: &str) -> Self {
        Self {
            whisper_model: whisper_model.to_string(),
            language: language.to_string(),
        }
    }
}

impl Default for DefaultAligner {
    fn default() -> Self {
        Self::new("base", "en")
    }
}

impl Aligner for DefaultAligner {
    fn name(&self) -> &str {
        "default"
    }

    fn process(
        &self,
        audio_path: &Path,
        model_dir: Option<&Path>,
    ) -> Result<AlignmentResult> {
        let result = transcribe::transcribe(
            audio_path,
            &self.whisper_model,
            &self.language,
            model_dir,
        )?;

        let syllables = syllabify::syllabify_words(&result.words);

        Ok(AlignmentResult {
            text: result.text,
            words: result.words,
            syllables,
        })
    }
}

/// Get an aligner backend by name.
///
/// Modes:
/// - "default" / "auto" — Whisper + G2P + ARPABET proportional timing.
/// - "bfa" — Not yet available natively (see issue #21).
pub fn get_aligner(
    name: &str,
    whisper_model: &str,
    language: &str,
    _device: &str,
) -> Result<Box<dyn Aligner>> {
    match name {
        "auto" | "default" => Ok(Box::new(DefaultAligner::new(whisper_model, language))),
        "bfa" => {
            bail!(
                "BFA aligner is not yet available in the native build. \
                 Use 'auto' or 'default' aligner instead. \
                 See https://github.com/A-U-Supply/glottisdale/issues/21"
            );
        }
        _ => bail!("Unknown aligner: '{}'. Available: default, auto", name),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_aligner_name() {
        let aligner = DefaultAligner::default();
        assert_eq!(aligner.name(), "default");
    }

    #[test]
    fn test_get_aligner_default() {
        let aligner = get_aligner("default", "base", "en", "cpu").unwrap();
        assert_eq!(aligner.name(), "default");
    }

    #[test]
    fn test_get_aligner_bfa() {
        let result = get_aligner("bfa", "base", "en", "cpu");
        match result {
            Err(e) => assert!(e.to_string().contains("not yet available"), "Error: {}", e),
            Ok(_) => panic!("Expected error for BFA aligner"),
        }
    }

    #[test]
    fn test_get_aligner_unknown() {
        let result = get_aligner("nonexistent", "base", "en", "cpu");
        assert!(result.is_err());
    }

    #[test]
    fn test_get_aligner_auto() {
        let aligner = get_aligner("auto", "base", "en", "cpu").unwrap();
        assert_eq!(aligner.name(), "default");
    }
}
